src/System.Collections/src/System/Collections/Generic/HashSet.cs

```diff
@@ -115,13 +115,45 @@ public HashSet(IEnumerable<T> collection, IEqualityComparer<T> comparer)
             var otherAsHashSet = collection as HashSet<T>;
             if (otherAsHashSet != null && AreEqualityComparersEqual(this, otherAsHashSet))
             {
-                _buckets = (int[])otherAsHashSet._buckets.Clone();
-                _slots = (Slot[])otherAsHashSet._slots.Clone();
+                int count = otherAsHashSet._count;
```

Comment:
Moving the hashset-source case out into another method has also made me realise the TrimExcess call is never needed in this case, either the path I've introduced or the existing will mean it's below the threshold for it.